### PR TITLE
Make it possible to build Ruby gem with system OpenSSL

### PR DESCRIPTION
**Pull request by stanhu**
This pull request adds a number of changes to enable compiling gRPC for s390x:

1. Bring back ability to compile against system OpenSSL

https://github.com/grpc/grpc/pull/23957 removed the ability to build against the system OpenSSL libraries. However, BoringSSL is not supported on such platforms as the IBM s390x. This commits restores the behavior of that merge request just for OpenSSL.

2. Make it possible to build Ruby gem with system OpenSSL

By installing the gem with `EMBED_OPENSSL=false`, the Ruby gem can be compiled using the system OpenSSL library. This is needed on systems where BoringSSL doesn't compile (e.g. s390x).

3. Allow configuration of number of jobs to build Ruby gem

Some systems may be memory-constrained, so launching N jobs for N processors will cause the compilation to run out of memory.  We add an environment variable, `MAKE_NPROC`, that allows users to use a custom number of jobs.

For example, I was able to build on s390x via:

```sh
MAKE_NPROC=1 EMBED_OPENSSL=false gem install pkg/grpc-1.42.0.dev.gem
```

Closes https://github.com/grpc/grpc/issues/15997

@veblush @jtattermusch

### CODE REVIEW

File: src/ruby/ext/grpc/extconf.rb [hunk 1]
```diff
@@ -65,13 +66,14 @@
...
 ENV['BUILDDIR'] = output_dir
+make_flags = use_embedded_ssl ? 'HAS_SYSTEM_OPENSSL_ALPN=true HAS_EMBEDDED_OPENSSL_ALPN=false' : ''
 
 unless windows
   puts 'Building internal gRPC into ' + grpc_lib_dir
   nproc = 4
   nproc = Etc.nprocessors * 2 if Etc.respond_to? :nprocessors
   make = bsd ? 'gmake' : 'make'
-  system("#{make} -j#{nproc} -C #{grpc_root} #{grpc_lib_dir}/libgrpc.a CONFIG=#{grpc_config} Q=")
+  system("#{make} -j#{nproc} -C #{grpc_root} #{grpc_lib_dir}/libgrpc.a CONFIG=#{grpc_config} Q= #{make_flags}")
```
Comment (stanhu) [inline src/ruby/ext/grpc/extconf.rb:76]: I think I got this reversed.

Comment (stanhu) [reply src/ruby/ext/grpc/extconf.rb:76]: These flags don't appear to be in `master`. I was testing against gRPC v1.30.2.

File: src/ruby/ext/grpc/extconf.rb [hunk 2]
```diff
@@ -94,6 +95,7 @@
 $LDFLAGS << ' -Wl,-wrap,memcpy' if linux
 $LDFLAGS << ' -static-libgcc -static-libstdc++' if linux
 $LDFLAGS << ' -static' if windows
+$LDFLAGS << ' -lssl' unless use_embedded_ssl
```
Comment (stanhu) [inline src/ruby/ext/grpc/extconf.rb:98]: This doesn't appear to work in `master` at the moment.

File: src/ruby/ext/grpc/extconf.rb (hunk 1 above)
Comment (stanhu) [reply src/ruby/ext/grpc/extconf.rb]: It looks like https://github.com/grpc/grpc/pull/23957 removed the ability to use the system OpenSSL. 😦

**Comment by jtattermusch** [comment]
Unless we have a very strong demand for this, I'm generally against any changes that make our build system harder to maintain and that increases the number of build configurations we support (and which we should test consequently).

In particular, we're trying to keep the complexity of our Makefile as low as possible, which is also one of the reasons why we introduced https://github.com/grpc/grpc/pull/23957 while ago.  Since s390x is not a platform we officially support, I don't think we're willing to take the maintenance burden of increasing our Makefile's complexity (e.g. by reintroducing support of building against OpenSSL). If anything, our Makefile should be getting simpler over time.

**Comment by stanhu** [comment]
@jtattermusch That's understandable, but note that the `CMakeLIsts.txt` still supports `gRPC_SSL_PROVIDER` and other flags that make it possible to build with an external libraries. Are you planning on removing that as well? I looked at using that for the Ruby gem, but the gem doesn't ship the `cmake` related files. I'm wondering if you would be open to a pull request that used `cmake`.

**Comment by jtattermusch** [comment]
> @jtattermusch That's understandable, but note that the `CMakeLists.txt` still supports `gRPC_SSL_PROVIDER` and other flags that make it possible to build with an external libraries. Are you planning on removing that as well? I looked at using that for the Ruby gem, but the gem doesn't ship the `cmake` related files. I'm wondering if you would be open to a pull request that used `cmake`.

we're planning to keep the option to build against openssl with cmake (out of necessity), but we don't plan to introduce that option for other build systems (the complexity argument above).

I think building ruby gem using cmake would be very difficult as the build system for ruby extensions uses make internally (which is one of the main reasons why we have to keep supporting make at least for project-internal use - otherwise we would have removed the make support long time ago). I think trying to revamp the grpc ruby extension build to use cmake is a lost cause and I recommend against trying that - you'd probably burn a LOT of your own time and it wouldn't lead to anywhere.

**Comment by stanhu** [comment]
@jtattermusch Ok, thanks. FYI, I did spend a few minutes trying to use `cmake` in `extconf.rb`. I got pretty far, except the one hiccup is that the current `CMakeLists.txt` tries to build almost everything (e.g. compiling some `.proto` files), whereas the current `Makefile` only builds `libgrpc.a`. There would need to be changes to `CMakeLists.txt` to whittle down the build list. Alternatively, the gem would have to include more of the source than it does.

I will mention that gems like `libgit2/rugged` [already requre `cmake`](https://github.com/libgit2/rugged/blob/b92520bbbee0d1f88f2e7ffb42cceb32e993ce1b/ext/rugged/extconf.rb#L92-L94) to build the gem, so supporting `cmake` isn't unprecedented.

I'd argue that providing the ability to use the system OpenSSL library is not just a architecture-specific issue. Having the ability to use the system OpenSSL library is important for [FIPS compliance](https://csrc.nist.gov/publications/detail/fips/140/2/final). I'd propose that the Ruby extension support something like a `--use-system-ssl` option to enable this.

**Pull request closed without merging**